reps/2023-01-09-aws-fleet-support.md

```diff
@@ -0,0 +1,138 @@
+
+### How EC2 fleet works?
+
+An EC2 Fleet contains configuration to launch a group of instances. There are [three types](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-request-type.html) of fleet requests, namely, `request`, `maintain` and `instant`. Because of the constraints mentioned in the **requirements** section (i.e., EC2 must not interfere with autoscaler activities), we would only be able to make use of `instant` type. Unlike, `maintain` and `request`, `instant` type `create_fleet` creates a fleet entity synchronously in an AWS account which is simply a configuration. Below are couple of observations related to `instant` type EC2 fleets:
+
+-   `instant` type fleets cannot be deleted without terminating instances it created.
+-   directly terminating instances created by `instant` fleet using `terminate_instances` API does not affect the fleet state.
+-   spot interruptions do not replace an instance created by `instant` fleet.
+-   there is no limit to number of EC2 `instant` type fleets one can have in an account.
+
+### How will EC2 fleets be created?
+
+An EC2 fleet configuration will be encapsulated within a single node type from the autoscaler perspective. We aim to create a new EC2 `instant` fleet as part of `create_node` method that is pretty much the entry point for any scale-up request from any other component (per diagram above). Ray's autoscaler will make sure `max_workers` config is honoured as well as perform any bin-packing of the resources across node types. EC2 automatically tags the instances created this way with a tag named `aws:ec2:fleet-id`.
+
+### How will the demand be converted into resources?
+
+Autoscaling monitor receives several demands from the metrics like vCPUs, memory, gpus etc. The [resource_demand_scheduler.py](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/resource_demand_scheduler.py) is responsible for converting the demand into number of workers per node type specified in the autoscaler config. However, it relies on the **node provider** to statically determine the CPU, memory, GPU and any custom resources. Hence, **node_provider** will determine CPU, memory and GPU from the `InstanceRequirements` and specified `InstanceType` parameters and aggregated them based on the latest family or high spec instance. Hence, autoscaler will end up spinning less nodes than necessary which avoid overscaling problem. However, the [current behavior](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/aws/node_provider.py#L403) also does not guarantee target capacity will reached. Autoscaling monitor loop will figure out any underscaling issue
```

Comment:
> aggregated them based on the latest family or high spec instance

To clarify, does this mean we consider the resource shape of a particular "node config" to be that of the largest instance in the group?

Reply:
For safety, it should probably be the smallest? +1 to clarifying either way.

Reply:
Yes, that's correct. @wuisawesome, any thoughts on why it should be smallest? My understanding is that when resource shape of largest instance is reported, it can take autoscaler multiple iterations to scale up appropriately if EC2 fleet ends up creating instances of smaller instance (in terms of cpu, memory or gpu) types. However other case will lead to overscaling, hence autoscaler has to terminate excess nodes.

Reply:
I'm wondering about the case where a fleet includes p3.2xlarge and p3.16xlarge, if someone has a `ray.remote(num_gpus=4)`, using the smallest would hang/inifinite loop on launching resources forever.

Reply:
Ah got it. In that case, we would want to show this error, correct?

```
No available node types can fulfill resource request {}. Add suitable node types to this cluster to resolve this issue.
```

Then I agree we should consider lowest spec instance. In that case, we may end up with nodes having more resources than necessary if EC2 fleet always spins up highest spec instance.

Reply:
yeah in the initial version of this, i think we'd want to scope this to only recommend people use instances of the same "resource shape" in the same node type (e.g. the spot/on-demand version of a node, or m6.Nxlarge, m5.Nxlarge, m4.Nxlarge) to avoid this problem and have that logic for safety. I think we'd need something that looks broader than the node provider interface to support different resource shapes well.

Reply:
agree, updated the rep. Also, I'd think part of the problem is bringing in predictability of what EC2 fleets is going to choose so that autoscaler can make accurate resource planning.

Reply:
Yup, +1 to recommend using instances of the same shape. If the user wants to use two different instance sizes, this could be handled as two different node config entries. In this case, the autoscaler would generally prefer to launch the larger node config if there is sufficient pending demand, but would otherwise launch the smaller node config.